### PR TITLE
Replace Elixir/ALE with Elixir Circuits

### PR DESCRIPTION
**Pull request by axelclark**
Update Home Weather Display example project
* Update Home Weather Display example to latest Nerves version
* Updated DHT port to 3, so it works on RPI0
* Created a project with mix nerves.new and added modules to it
* Updated project to enable testing of grovepi update to Elixir Circuits

Replace ElixirAle with Elixir Circuits
* Replace ElixirAle.I2C with Circuits.I2C
* Elixir Circuits is the new I2C package from the ElixirAle maintainer
* Circuits uses NIFs instead of Ports
----
This is still a work in progress.  The update works with the LCD screen and DHT in the Home Weather Display project on actual hardware.

I converted `GrovePi.Board` to a `GenServer` to hold the `I2C ref`.  See [porting instructions](https://github.com/elixir-circuits/circuits_i2c/blob/master/PORTING.md#porting-elixirale-to-circuitsi2c).

I haven't figured out the best way to update the `GrovePi.I2C` and tests.  One approach I've considered is to add all of the test-specific `GrovePi.I2C` functions (e.g. `get_last_write/1`) to the `GrovePi.Board` then send and request the data from the `GrovePi.I2C` GenServer started when `GrovePi.Board` calls `@i2c.open/1` in `handle_continue/2`.

### CODE REVIEW

**Comment by axelclark** [comment]
I added the test helper functions to `GrovePi.Board` then sent request to `I2C`.  Had to go through `GrovePi.Board` because it holds the pid/ref for `I2C` in its state.

Remaining failing test is because retries now happen within Circuits rather than GrovePi.  When the poller receives an error from the I2C board, should it fail (current implementation) or somehow handle the error?

https://github.com/adkron/grovepi/blob/f76df053cdfcbad7cff887414a751178c8c5a3d4/lib/grovepi/poller.ex#L168-L173

Current test has 1 error which is discarded during the 1st retry in `GrovePi.Board`.

https://github.com/adkron/grovepi/blob/f76df053cdfcbad7cff887414a751178c8c5a3d4/test/grovepi/ir_reflective_test.exs#L20-L30

https://github.com/adkron/grovepi/blob/f76df053cdfcbad7cff887414a751178c8c5a3d4/lib/grovepi/board.ex#L78-L85

Here is the new implementation:
```elixir
  def handle_call({:read, bytes_to_read}, _from, state) do
    reply =
      case(@i2c.read(state.i2c_bus, state.address, bytes_to_read, retries: @i2c_retry_count)) do
        {:ok, response} -> response
        {:error, error} -> {:error, error}
      end


    {:reply, reply, state}
  end
```

**Comment by axelclark** [comment]
This was my best attempt at updating, but I welcome any feedback for a better implementation.

File: lib/grovepi/board.ex [hunk 1]
```diff
@@ -56,40 +60,112 @@ defmodule GrovePi.Board do
...
+  def handle_call({:write_device, address, message}, _from, state) do
+    reply = @i2c.write(state.i2c_bus, address, message, retries: @i2c_retry_count)
+    {:reply, reply, state}
+  end
+
+  @impl true
+  def handle_call({:read, bytes_to_read}, _from, state) do
+    reply =
+      case(@i2c.read(state.i2c_bus, state.address, bytes_to_read, retries: @i2c_retry_count)) do
+        {:ok, response} -> response
```
Comment (fhunleth) [inline lib/grovepi/board.ex:139]: I think that you want to keep the `{:ok, _}` part of the tuple here so that callers can more easily detect success.

Comment (fhunleth) [reply lib/grovepi/board.ex:139]: Looks like this was the old behavior, so I take back my comment for this PR.

**Comment by adkron** [comment]
I’ve been offline the last few days. This is on my to do for tomorrow and the next day. Thanks. 

Amos King
CEO
Binary Noggin

**Comment by adkron** [comment]
I think if there is an error from the i2c we can let the poller fail and restart. If we can come back from the failure, I believe that is best because we might not know the state of why, and a restart is probably a good idea.

I think we can always change our minds on that and go from failing to error handling later, but it is hard to go the other way once people start to depend on the implementation.

File: lib/grovepi/board.ex [hunk 2]
```diff
@@ -56,40 +60,112 @@ defmodule GrovePi.Board do
...
-      response -> response
-    end
+  @impl true
+  def init(state) do
+    {:ok, state, {:continue, []}}
   end
 
-  defp get_response_with_retry(_board, _len, 0), do: {:error, :too_many_retries}
+  @impl true
+  def handle_continue(_continue, state) do
```
Comment (adkron) [inline lib/grovepi/board.ex]:
```suggestion
  def handle_continue(:init, state) do
```

File: lib/grovepi/board.ex [hunk 3]
```diff
@@ -56,40 +60,112 @@ defmodule GrovePi.Board do
...
+  ## Server Callbacks
 
-  defp send_request_with_retry(board, message, retries_left) do
-    case @i2c.write(board, message) do
-      {:error, _} -> send_request_with_retry(board, message, retries_left - 1)
-      response -> response
-    end
+  @impl true
+  def init(state) do
+    {:ok, state, {:continue, []}}
```
Comment (adkron) [inline lib/grovepi/board.ex]:
```suggestion
    {:ok, state, {:continue, :init}}
```

Comment (adkron) [reply lib/grovepi/board.ex]: We should have something that designates what the continue is. At the moment I think saying what we are continuing is probably the best option.

File: lib/grovepi/board.ex (hunk 2 above)
Comment (axelclark) [reply lib/grovepi/board.ex]: How about `:open_i2c` instead of `:init`?

File: lib/grovepi/board.ex (hunk 3 above)
Comment (axelclark) [reply lib/grovepi/board.ex]: How about `:open_i2c` instead of `:init`?

Comment (adkron) [reply lib/grovepi/board.ex]: That is even better.

Comment (axelclark) [reply lib/grovepi/board.ex]: Great suggestion, it's much clearer why there is a continue now.

**Comment by axelclark** [comment]
I made the updates based on your comments.  

One additional note.  The compiler sends warning messages for the `GrovePi.Board` test helper callbacks when compiling for prod because those `@i2c` functions don't exist in Circuits.I2C.  I'm not sure how to remove those warnings, change the implementation to eliminate them, or if those should be ignored.  For example:
```elixir
  def handle_call({:get_last_write}, _from, state) do
    {:reply, @i2c.get_last_write(state.i2c_bus), state}
  end
```

Let me know if you see anything else.  Thanks for the feedback and the opportunity to contribute to the project!

**Comment by adkron** [comment]
I might need to rework the testing and how we are doing that part of the project. I think we can pull this in with the warnings and work on those later. Moving to Elixir Circuits is important work. Thank you very much. 🤖

**Pull request merged**